### ISSUE TITLE
[HOLD for payment 2023-07-21] [$1000] Share somewhere doesn't update in room while selecting user/group

### ISSUE THREAD

**Issue by kavimuru** [Expensify/App#21614]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

## Action Performed:
1. Create a room
2. select '+" icon and click assign task ( It must be done within room)
3. Add title and description
4. Go to share somewhere and click on any other rooms/ contact list.
( Notice that while clicking other groups too, only default room on which the task is created is shown.)
## Expected Result: 
Either, multiple groups should be selectable and should be updated as like assign task done from LHN,
OR, sharing somewhere should be made unchooseable, and only default room should be shown
## Actual Result: 
Share somewhere doesn't update in room while selecting user/group

## Workaround:
Can the user still use Expensify without this being fixed? Have you informed them of the workaround?

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android / native
- [ ] Android / Chrome
- [ ] iOS / native
- [ ] iOS / Safari
- [x] MacOS / Chrome / Safari
- [ ] MacOS / Desktop

**Version Number:** 1.3.32-5
**Reproducible in staging?:** y
**Reproducible in production?:** y
**If this was caught during regression testing, add the test name, ID and link from TestRail:**
**Email or phone of affected tester (no customers):**
**Logs:** https://stackoverflow.com/c/expensify/questions/4856
**Notes/Photos/Videos:** Any additional supporting documentation

https://github.com/Expensify/App/assets/43996225/c7b7b9bb-10fd-4e8a-93c7-bfe4b49cba87

https://github.com/Expensify/App/assets/43996225/d64cff13-382b-4039-ab2a-1ea45a866eb2

**Expensify/Expensify Issue URL:**
**Issue reported by:** @ashimsharma10
**Slack conversation:** https://expensify.slack.com/archives/C049HHMV9SM/p1687351420861259

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01cdc2396412f39341</li>
        <li>Upwork Job ID: 1674191448143474688</li>
        <li>Last Price Increase: 2023-07-05</li>
    </ul>
</details>

**Comment by StevenKKC** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Share somewhere doesn't update in room while selecting user/group.

### What is the root cause of that problem?
When select user/group in share somewhere page, `task.shareDestination` will be set.
After return to create task page, because `task.parentReportID` is set, `task.shareDestination` will be set to original value.

Therefor share somewhere doesn't update.

### What changes do you think we should make in order to solve the problem?
When select user/group, we should clear `task.parentReportID`.

### What alternative solutions did you explore? (Optional)
None.

**Comment by Nikhil-Vats** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Share somewhere doesn't update when creating a task from a chat (report).

### What is the root cause of that problem?
In `NewTaskPage` we override the share somewhere in the lines below so it cannot be changed to any user/group/room other than the parent report if we create a task from a report.
https://github.com/Expensify/App/blob/67561d81bd95681e47dab14261349eb6eaad384d/src/pages/tasks/NewTaskPage.js#L95-L100

### What changes do you think we should make in order to solve the problem?
Since, the user has clicked on assign task from a chat and not FAB in LHN, we should disable the share somewhere field when creating a task from a report.

This is also consistent with send/request money in a report. We do not let user change the payee in request/send money page if they opened the page from a report.

To disable share somewhere when we click on assign task from a report, we should -
```javascript
<MenuItem
    disabled={Boolean(props.task.parentReportID)}
    label={shareDestination.displayName ? props.translate('newTaskPage.shareSomewhere') : ''}
    title={shareDestination.displayName || ''}
    description={shareDestination.displayName ? shareDestination.subtitle : props.translate('newTaskPage.shareSomewhere')}
    icon={shareDestination.icons}
    onPress={() => Navigation.navigate(ROUTES.NEW_TASK_SHARE_DESTINATION)}
    shouldShowRightIcon
/>
```

Additionally, we can also return from the `onPress` function if user is creating a task from a report.
```javascript
onPress={() => {
    if(props.task.parentReportID) return;
    Navigation.navigate(ROUTES.NEW_TASK_SHARE_DESTINATION)
}}
```
### What alternative solutions did you explore? (Optional)
NA

**Comment by Nikhil-Vats** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
Share somewhere doesn't update when creating a task from a chat (report).

### What is the root cause of that problem?
In `TaskShareDestinationSelectorModal`, we don't update `parentReportID` when updating share destination -
https://github.com/Expensify/App/blob/67561d81bd95681e47dab14261349eb6eaad384d/src/pages/tasks/TaskShareDestinationSelectorModal.js#L117-L122

In `NewTaskPage` we override the share somewhere in the lines below so it cannot be changed to any user/group/room other than the parent report if we create a task from a report.
https://github.com/Expensify/App/blob/67561d81bd95681e47dab14261349eb6eaad384d/src/pages/tasks/NewTaskPage.js#L95-L100

### What changes do you think we should make in order to solve the problem?
We should also change the `parentReportID` when changing the share somewhere destination in `TaskShareDestinationSelectorModal` -

To do that, we can call the `setParentReportID` in `setShareDestinationValue` function in `TaskUtils` -
```javascript
/**
 * Sets the shareDestination value for the task
 * @param {string} shareDestination
 */
function setShareDestinationValue(shareDestination) {
    // This is only needed for creation of a new task and so it should only be stored locally
    Onyx.merge(ONYXKEYS.TASK, {shareDestination});
    setParentReportID(shareDestination); // call this function also when changing the share destination
    // this will also update the parentReportID so that share destination will not be overridden in `NewTaskPage`
}
```
`setParentReportID` is defined after `setShareDestinationValue` so we will need to reorder and define `setParentReportID` before `setShareDestinationValue` to use it in `setShareDestinationValue`.

Alternatively, we can export `setParentReportID` and use it in `TaskShareDestinationSelectorModal` -
```javascript
if (option.reportID) {
    // Clear out the state value, set the assignee and navigate back to the NewTaskPage
    setSearchValue('');
    TaskUtils.setShareDestinationValue(option.reportID); // alternatively, call `setParentReportID` in this function and remove next line
    TaskUtils.setParentReportID(option.reportID);
    Navigation.goBack();
}
```
### What alternative solutions did you explore? (Optional)
Since, the user has clicked on assign task from a chat and not FAB in LHN, we can disable the share somewhere field when creating a task from a report.

This is consistent with send/request money in a report. We do not let user change the payee in request/send money page if they opened the page from a report.

To disable share somewhere when we click on assign task from a report, we can add the disabled prop to the field if parentReportID is set -
```javascript
<MenuItem
    disabled={Boolean(props.task.parentReportID)} // disable the field if parentReportID is set
    label={shareDestination.displayName ? props.translate('newTaskPage.shareSomewhere') : ''}
    title={shareDestination.displayName || ''}
    description={shareDestination.displayName ? shareDestination.subtitle : props.translate('newTaskPage.shareSomewhere')}
    icon={shareDestination.icons}
    onPress={() => Navigation.navigate(ROUTES.NEW_TASK_SHARE_DESTINATION)}
    shouldShowRightIcon
/>
```

Additionally, we can also return from the `onPress` function if user is creating a task from a report.
```javascript
onPress={() => {
    if(props.task.parentReportID) return;
    Navigation.navigate(ROUTES.NEW_TASK_SHARE_DESTINATION)
}}
```

**Comment by bernhardoj** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.
We can't change the Share Somewhere field when creating a task within a report.

### What is the root cause of that problem?
When we are creating a task within a report, it will autofill the Share Somewhere field (currently it's not working because of this [issue](https://github.com/Expensify/App/issues/21361)) with the report we initiated the new task on.
https://github.com/Expensify/App/blob/e5f00f86fb412702e9c212b5a88766a8ef7da611/src/pages/tasks/NewTaskPage.js#L95-L100

`parentReportID` is the ID of the report on which we initiated the new task.

The problem is, the code above is put inside a `useEffect` with a dependency array of `[props]` which means it will rerun every time ANY props are changed. This will produce unexpected side effects as we have in this issue.

In this case, when we select a new Share Somewhere, it will also trigger the code above which will just override the new selected Share Somewhere with the `parentReportID`.

### What changes do you think we should make in order to solve the problem?

To prevent any unexpected side effects, we should extract it out into a separate `useEffect` with a dependency array that is only needed, in this case, `props.task.parentReportID`.
```
useEffect(() => {
    if (props.task.parentReportID) {
        TaskUtils.setShareDestinationValue(props.task.parentReportID);
    }
}, [props.task.parentReportID]);
```

---
We can do the same thing to all (4) conditions inside the `useEffect`.
https://github.com/Expensify/App/blob/e5f00f86fb412702e9c212b5a88766a8ef7da611/src/pages/tasks/NewTaskPage.js#L75-L109

**Comment by alexpensify** [comment]
Adding the `external` label because I'm able to reproduce. @sobitneupane to keep us moving forward, can you review if any of these proposals will correct the issue? Thanks! 

![2023-06-28_16-00-12](https://github.com/Expensify/App/assets/8107070/64e7a552-9214-407f-9db2-e805ed8f1eea)

![2023-06-28_16-00-03](https://github.com/Expensify/App/assets/8107070/b78f9997-a3ab-4441-8d0c-8dd22df44932)

**Comment by thienlnam** [comment]
This is a duplicate of https://github.com/Expensify/App/issues/20898

**Comment by alexpensify** [comment]
Whoops I missed that one when checked GH. I'm closing and I closed the job in upwork too

**Comment by bernhardoj** [comment]
Looking into the root cause of https://github.com/Expensify/App/issues/20898, I think the issue is different. https://github.com/Expensify/App/issues/20898 issue is about selecting a chat that doesn't exist and the solution is to only show chat/report that exist in the list.

In this issue, we select an exist chat/report, but then the value is overwritten because of an unexpected side effect of `useEffect`

**Comment by ashimsharma10** [comment]
I second @bernhardoj . in this issue we are talking about the selected user not being saved.

**Comment by thienlnam** [comment]
Thanks for the comment - since they were both regarding a selection in share somewhere not working I assumed it was the same - will open this back up

**Comment by alexpensify** [comment]
Alright, @sobitneupane when you get a chance, can you please review the proposal? Thank you!

**Comment by alexpensify** [comment]
Reassigning another `bug` team member, I'm going OOO until Monday, July 10, and will take it back then if it's not closed by then. 

Required action from the 🐛 team: Make sure the proposal process is moving forward, so we can get this one assigned out to fix.

**Comment by thienlnam** [comment]
cc @sobitneupane Any update here?

**Comment by sobitneupane** [comment]
@thienlnam Can you please confirm if we want to allow user to change "Share somewhere" when it is opened from a report (not FAB). There are two possible solutions here. One is to disable "Share somewhere" when task assignment begins from a report. Other is to allow change in "Share somewhere".

**Comment by thienlnam** [comment]
@sobitneupane Let's disable the "Share somewhere" option like we originally intended. If you're creating a task within a specific chat - chances are you already are aware the location to share it would not change

**Comment by sobitneupane** [comment]
Thanks @thienlnam. Following image shows the UI after disabling the `MenuItem` as suggested by @Nikhil-Vats in his [alternative solution](https://github.com/Expensify/App/issues/21614#issuecomment-1608287127). Can you please confirm if it is the expected UI? 

<img width="456" alt="Screenshot 2023-07-07 at 15 52 11" src="https://github.com/Expensify/App/assets/25876548/cca7b7f5-620a-4efb-bbaf-f434d82c78b5">

**Comment by Nikhil-Vats** [comment]
@sobitneupane If we don't want to show it disabled, then we can just disable click behaviour by returning early from `onPress` function and show disabled cursor -

https://github.com/Expensify/App/assets/31413407/7c872f0e-cd04-4d51-91d7-3ae01404dddf

@thienlnam let us know which one you prefer more.

**Comment by sobitneupane** [comment]
> If we don't want to show it disabled, then we can just disable click behaviour by returning early from onPress function and show disabled cursor -

I don't think it's a good idea. If we don't want to show it disabled, we can do something similar to `ReportDetailsPage`.
For instance:
```
{shouldAllowWriteCapabilityEditing ? (
                            <MenuItemWithTopDescription
                                shouldShowRightIcon
                                title={writeCapabilityText}
                                description={this.props.translate('writeCapabilityPage.label')}
                                onPress={() => Navigation.navigate(ROUTES.getReportSettingsWriteCapabilityRoute(this.props.report.reportID))}
                            />
                        ) : (
                            <View style={[styles.ph5, styles.pv3]}>
                                <Text
                                    style={[styles.textLabelSupporting, styles.lh16, styles.mb1]}
                                    numberOfLines={1}
                                >
                                    {this.props.translate('writeCapabilityPage.label')}
                                </Text>
                                <Text
                                    numberOfLines={1}
                                    style={[styles.optionAlternateText, styles.pre]}
                                >
                                    {writeCapabilityText}
                                </Text>
                            </View>
                        )}
```

**Comment by Nikhil-Vats** [comment]
@sobitneupane What do you think is not a good idea - an early return from `onPress` or the `disabled` cursor or both?

We can disable interaction like you suggested or use the props like @bernhardoj suggested above. I added the property for `disabled` cursor for consistency as we also do it for split bill flow -

https://github.com/Expensify/App/assets/31413407/4e0ae572-f604-4929-be1d-5e5da817f0d9

**Comment by thienlnam** [comment]
![Screenshot 2023-07-07 at 10 21 45 AM](https://github.com/Expensify/App/assets/30609178/4730326f-5539-42f4-b068-ed8d9818882a)
We should definitely use this UI - it's more consistent with the disabled push to page inputs we have already

**Comment by Nikhil-Vats** [comment]
Sure @thienlnam and what about the cursor, should it be disabled, or should it be just the default cursor? We have disabled cursor here -

https://github.com/Expensify/App/assets/31413407/4e0ae572-f604-4929-be1d-5e5da817f0d9

**Comment by thienlnam** [comment]
We can copy that pattern and use the disabled cursor on highlight

**Comment by Nikhil-Vats** [comment]
@sobitneupane Updated the [proposal](https://github.com/Expensify/App/issues/21614#issuecomment-1608287127) to follow the suggestions by you and Jack.

**Comment by sobitneupane** [comment]
Thanks @thienlnam.

[Solution](https://github.com/Expensify/App/issues/21614#issuecomment-1625227872) proposed by @bernhardoj looks good to me. 

🎀 👀 🎀 C+ reviewed

**Comment by thienlnam** [comment]
Will take BZE on this issue since I've already been involved in it

**Comment by alexpensify** [comment]
Alright, removing the other 🐛 team member. I'm back online and see that this is our current state: 

Waiting for a PR from @bernhardoj

**Comment by ashimsharma10** [comment]
My upwork profile is : https://www.upwork.com/freelancers/~018a92cf13e1e88eed .

Idk why its needed again.

**Comment by alexpensify** [comment]
Great work team, now we will wait for automation here.

**Comment by alexpensify** [comment]
Still waiting for automation

**Comment by alexpensify** [comment]
@ashimsharma10 and @bernhardoj - I've sent the payment request via Upwork. I had to use a new job link, please accept and I can complete the required process. 

--- 

@sobitneupane - Please complete the checklist and request payment via Chat. 

--- 

Thank you!

**Comment by alexpensify** [comment]
@ashimsharma10 and @bernhardoj - I've completed the payment process via Upwork! 

--- 

@sobitneupane - please complete the checklist, so we can continue with the next steps for payment here and we can close the GH out. Thank you!

**Comment by sobitneupane** [comment]
> **BugZero Checklist:** The PR fixing this issue has been merged! The following checklist ([instructions](https://stackoverflow.com/c/expensify/questions/14574)) will need to be completed before the issue can be closed:
> 
> * [ ]  [@sobitneupane] The PR that introduced the bug has been identified. Link to the PR:

https://github.com/Expensify/App/pull/20145
> * [ ]  [@sobitneupane] The offending PR has been commented on, pointing out the bug it caused and why, so the author and reviewers can learn from the mistake. Link to comment:

https://github.com/Expensify/App/pull/20145/files#r1272428756
> * [ ]  [@sobitneupane] A discussion in #expensify-bugs has been started about whether any other steps should be taken (e.g. updating the PR review checklist) in order to catch this type of bug sooner. Link to discussion:

This should have been caught in PR review. No change in PR review checklist required.

> * [ ]  [@sobitneupane] Determine if we should create a regression test for this bug.

Yes.
> * [ ]  [@sobitneupane] If we decide to create a regression test for the bug, please propose the [regression test](https://github.com/Expensify/App/blob/main/contributingGuides/REGRESSION_TEST_BEST_PRACTICES.md) steps to ensure the same bug will not reach production again.

**Comment by sobitneupane** [comment]
**Regression Test Proposal**

1. Open any chat
2. Press the + button in Composer
3. Press Assign Task
4. Enter any title and press Next
5. Verify Share somewhere doesn't have a right arrow
6. Verify pressing/clicking Share somewhere doesn't do anything
7. Now, go back to LHN and press the FAB + button
8. Repeat steps 3-4
9. Verify Share somewhere has a right arrow
10. Verify pressing/clicking Share somewhere open a new page to select a report

Do we agree 👍 or 👎

**Comment by sobitneupane** [comment]
Requested payment on newDot.

**Comment by alexpensify** [comment]
@sobitneupane, thank you! @thienlnam can you please review the checklist and give a 👍🏼? Thanks.

**Comment by thienlnam** [comment]
Thumbbed it, looks good to me

**Comment by alexpensify** [comment]
Thanks, I'm going to close this one out!

**Comment by alexpensify** [comment]
Here is the payment summary: 

-  External issue reporter: @ashimsharma10 $250
-  Contributor that fixed the issue: @bernhardoj $1500 
-  Contributor+ that helped on the issue and/or PR: @sobitneupane $1500

Upwork Job: https://www.upwork.com/jobs/~01c3b0b1ecadacf47b

*If applicable, the bonuses will be applied on the final payment 

Extra Notes regarding payment: https://github.com/Expensify/App/issues/20320#issuecomment-1642715820

All set here @JmillsExpensify! This one was pre-summary days.

**Comment by JmillsExpensify** [comment]
Reviewed details for @sobitneupane. This is accurate based on summary from Business Reviewer and approved for payment in NewDot.
